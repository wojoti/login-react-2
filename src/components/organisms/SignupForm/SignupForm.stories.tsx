/* eslint-disable react/function-component-definition */
/* eslint-disable react/jsx-props-no-spreading */
/* eslint-disable import/no-extraneous-dependencies */
import { ComponentMeta, ComponentStory } from "@storybook/react";

import SignupForm from "./SignupForm";

// More on default export: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
export default {
  title: "Organisms/SignupForm",
  component: SignupForm,
} as ComponentMeta<typeof SignupForm>;

// More on component templates: https://storybook.js.org/docs/react/writing-stories/introduction#using-args
const Template: ComponentStory<typeof SignupForm> = (args) => (
  <SignupForm {...args} />
);

export const Default = Template.bind({});
Default.args = {};
