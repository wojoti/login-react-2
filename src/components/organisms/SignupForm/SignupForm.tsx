import Button from "@components/atoms/Button/Button";
import Column from "@components/atoms/Column/Column";
import TextInput from "@components/molecules/TextInput/TextInput";
import { forwardRef, useImperativeHandle, useRef } from "react";

type SignupFormProps = {
  onSignupSubmit: (email: string, password: string) => void;
};

export type SignupFormHandle = {
  focus: () => void;
};
const SignupForm = forwardRef<SignupFormHandle, SignupFormProps>(
  ({ onSignupSubmit }, ref) => {
    const emailRef = useRef<HTMLInputElement>(null);
    const passwordRef = useRef<HTMLInputElement>(null);
    useImperativeHandle(
      ref,
      () => ({
        focus: () => {
          emailRef.current?.focus();
        },
      }),
      []
    );
    const onSubmit = () => {
      onSignupSubmit(emailRef.current!.value, passwordRef.current!.value);
    };
    return (
      <>
        <Column mt={15}>
          <TextInput type="email" name="Email" id="email" ref={emailRef} />
        </Column>
        <Column mt={15}>
          <TextInput
            type="password"
            name="Password"
            id="password"
            ref={passwordRef}
          />
        </Column>
        <Column mt={15}>
          <Button name="SIGN UP" onClick={onSubmit} testId="submit-button-id" />
        </Column>
      </>
    );
  }
);
SignupForm.displayName = "SignupForm";

export default SignupForm;
