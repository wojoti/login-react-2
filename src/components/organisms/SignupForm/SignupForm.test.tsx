import { fireEvent, render, screen, within } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import TestRenderer from "react-test-renderer";
import SignupForm, { SignupFormProps } from "./SignupForm";

const props: SignupFormProps = {
  onSignupSubmit: () => {},
  testId: "test-signupform-id",
};

test("should match snapshot", () => {
  const tree = TestRenderer.create(<SignupForm {...props} />).toJSON();
  expect(tree).toMatchSnapshot();
});

test("should render signupform", () => {
  render(<SignupForm {...props} />);
  const signupformElement = screen.getByTestId("test-signupform-id");
  expect(signupformElement).toBeInTheDocument();
});

test("should render signupform default disabled button ", () => {
  render(<SignupForm {...props} />);
  const signupformElement = screen.getByTestId("test-signupform-id");
  expect(signupformElement).toBeInTheDocument();
  const signupformButtonElement = screen.getByTestId("signup-button-id");
  expect(signupformButtonElement).toBeInTheDocument();
  expect(signupformButtonElement).toBeDisabled();
});

test("should render signupform with enabled button - both inputs filled, email regex correct", () => {
  render(<SignupForm {...props} />);
  const signupformElement = screen.getByTestId("test-signupform-id");
  expect(signupformElement).toBeInTheDocument();
  const signupformLoginElement = screen.getByTestId(
    "signupform-textinput-login"
  );
  expect(signupformLoginElement).toBeInTheDocument();
  const textfieldLoginElement = within(signupformLoginElement).getByTestId(
    "textinput-textfield"
  );
  expect(textfieldLoginElement).toBeInTheDocument();
  userEvent.type(textfieldLoginElement, "test@test.eu");

  const signupformPasswordElement = screen.getByTestId(
    "signupform-textinput-password"
  );
  expect(signupformPasswordElement).toBeInTheDocument();
  const textfieldPasswordElement = within(
    signupformPasswordElement
  ).getByTestId("textinput-textfield");
  expect(textfieldPasswordElement).toBeInTheDocument();
  userEvent.type(textfieldPasswordElement, "pwd");

  const signupformButtonElement = screen.getByTestId("signup-button-id");
  expect(signupformButtonElement).toBeInTheDocument();
  expect(signupformButtonElement).not.toBeDisabled();
});

test("should handle onClick on button", () => {
  const handleClick = jest.fn();
  render(<SignupForm {...props} onSignupSubmit={handleClick} />);
  const signupformElement = screen.getByTestId("test-signupform-id");
  expect(signupformElement).toBeInTheDocument();
  const signupformLoginElement = screen.getByTestId(
    "signupform-textinput-login"
  );
  expect(signupformLoginElement).toBeInTheDocument();
  const textfieldLoginElement = within(signupformLoginElement).getByTestId(
    "textinput-textfield"
  );
  expect(textfieldLoginElement).toBeInTheDocument();
  userEvent.type(textfieldLoginElement, "test@test.eu");

  const signupformPasswordElement = screen.getByTestId(
    "signupform-textinput-password"
  );
  expect(signupformPasswordElement).toBeInTheDocument();
  const textfieldPasswordElement = within(
    signupformPasswordElement
  ).getByTestId("textinput-textfield");
  expect(textfieldPasswordElement).toBeInTheDocument();
  userEvent.type(textfieldPasswordElement, "pwd");

  const signupformButtonElement = screen.getByTestId("signup-button-id");
  expect(signupformButtonElement).toBeInTheDocument();
  expect(signupformButtonElement).not.toBeDisabled();

  fireEvent.click(signupformButtonElement);
  expect(handleClick).toHaveBeenCalledTimes(1);
});

test("should render signupform with disabled button - both inputs filled, email regex incorrect", () => {
  render(<SignupForm {...props} />);
  const signupformElement = screen.getByTestId("test-signupform-id");
  expect(signupformElement).toBeInTheDocument();
  const signupformLoginElement = screen.getByTestId(
    "signupform-textinput-login"
  );
  expect(signupformLoginElement).toBeInTheDocument();
  const textfieldLoginElement = within(signupformLoginElement).getByTestId(
    "textinput-textfield"
  );
  expect(textfieldLoginElement).toBeInTheDocument();
  userEvent.type(textfieldLoginElement, "t");

  const signupformPasswordElement = screen.getByTestId(
    "signupform-textinput-password"
  );
  expect(signupformPasswordElement).toBeInTheDocument();
  const textfieldPasswordElement = within(
    signupformPasswordElement
  ).getByTestId("textinput-textfield");
  expect(textfieldPasswordElement).toBeInTheDocument();
  userEvent.type(textfieldPasswordElement, "pwd");

  const signupformButtonElement = screen.getByTestId("signup-button-id");
  expect(signupformButtonElement).toBeInTheDocument();
  expect(signupformButtonElement).toBeDisabled();
});

test("should render signupform with disabled button - login filled, email regex correct", () => {
  render(<SignupForm {...props} />);
  const signupformElement = screen.getByTestId("test-signupform-id");
  expect(signupformElement).toBeInTheDocument();
  const signupformLoginElement = screen.getByTestId(
    "signupform-textinput-login"
  );
  expect(signupformLoginElement).toBeInTheDocument();
  const textfieldLoginElement = within(signupformLoginElement).getByTestId(
    "textinput-textfield"
  );
  expect(textfieldLoginElement).toBeInTheDocument();
  userEvent.type(textfieldLoginElement, "test@test.eu");

  const signupformButtonElement = screen.getByTestId("signup-button-id");
  expect(signupformButtonElement).toBeInTheDocument();
  expect(signupformButtonElement).toBeDisabled();
});

test("should render signupform with disabled button - password filled", () => {
  render(<SignupForm {...props} />);
  const signupformElement = screen.getByTestId("test-signupform-id");
  expect(signupformElement).toBeInTheDocument();
  const signupformPasswordElement = screen.getByTestId(
    "signupform-textinput-password"
  );
  expect(signupformPasswordElement).toBeInTheDocument();
  const textfieldPasswordElement = within(
    signupformPasswordElement
  ).getByTestId("textinput-textfield");
  expect(textfieldPasswordElement).toBeInTheDocument();
  userEvent.type(textfieldPasswordElement, "pwd");

  const signupformButtonElement = screen.getByTestId("signup-button-id");
  expect(signupformButtonElement).toBeInTheDocument();
  expect(signupformButtonElement).toBeDisabled();
});
test("should render email textfield with red border when invalid (eg. regex)", () => {
  render(<SignupForm {...props} />);
  const signupformElement = screen.getByTestId("test-signupform-id");
  expect(signupformElement).toBeInTheDocument();
  const signupformLoginElement = screen.getByTestId(
    "signupform-textinput-login"
  );
  expect(signupformLoginElement).toBeInTheDocument();
  const textfieldLoginElement = within(signupformLoginElement).getByTestId(
    "textinput-textfield"
  );
  expect(textfieldLoginElement).toBeInTheDocument();
  userEvent.type(textfieldLoginElement, "t");
  expect(textfieldLoginElement).toHaveStyleRule(
    "border",
    "2px solid rgb(244 63 94)"
  );
});

test("should not render password textfield with red border when invalid'", () => {
  render(<SignupForm {...props} />);
  const signupformElement = screen.getByTestId("test-signupform-id");
  expect(signupformElement).toBeInTheDocument();
  const signupformLoginElement = screen.getByTestId(
    "signupform-textinput-password"
  );
  expect(signupformLoginElement).toBeInTheDocument();
  const textfieldLoginElement = within(signupformLoginElement).getByTestId(
    "textinput-textfield"
  );
  expect(textfieldLoginElement).toBeInTheDocument();
  userEvent.type(textfieldLoginElement, "t");
  expect(textfieldLoginElement).not.toHaveStyleRule(
    "border",
    "2px solid rgb(244 63 94)"
  );
});
