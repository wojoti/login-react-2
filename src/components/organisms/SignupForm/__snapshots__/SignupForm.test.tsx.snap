// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
.c5 {
  box-sizing: border-box;
  margin: 0;
  padding: 0;
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI","Roboto","Oxygen", "Ubuntu","Cantarell","Fira Sans","Droid Sans","Helvetica Neue", sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  width: 100%;
  height: 32px;
  border-radius: 6px;
  border: 0px solid black;
  background-color: rgb(244 63 94);
  background-color: rgb(244 63 94);
  color: white;
  cursor: pointer;
  box-shadow: 0 25px 50px -12px rgb(0 0 0 / 0.25);
}

.c1 {
  box-sizing: border-box;
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI","Roboto","Oxygen", "Ubuntu","Cantarell","Fira Sans","Droid Sans","Helvetica Neue", sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  margin-top: 15px;
  margin-bottom: 0px;
  margin-left: 0px;
  margin-right: 0px;
  padding-top: 0px;
  padding-bottom: 0px;
  padding-left: 0px;
  padding-right: 0px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.c3 {
  box-sizing: border-box;
  margin: 0;
  padding: 0;
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI","Roboto","Oxygen", "Ubuntu","Cantarell","Fira Sans","Droid Sans","Helvetica Neue", sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  color: inherit;
  font-size: 14px;
}

.c4 {
  box-sizing: border-box;
  margin: 0;
  padding: 0;
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI","Roboto","Oxygen", "Ubuntu","Cantarell","Fira Sans","Droid Sans","Helvetica Neue", sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  width: 100%;
  height: 32px;
  border: 2px solid #e5e7eb;
  border-radius: 5px;
  padding-left: 10px;
}

.c2 {
  width: 100%;
}

.c0 {
  width: 100%;
}

<div
  className="c0"
  data-testid="test-signupform-id"
>
  <div
    className="c1"
  >
    <div
      className="c2"
      data-testid="signupform-textinput-login"
    >
      <span
        className="c3"
        data-testid="textinput-label"
      >
        Email
      </span>
      <input
        className="c4"
        data-testid="textinput-textfield"
        id="email"
        name="Email"
        onChange={[Function]}
        type="email"
      />
    </div>
  </div>
  <div
    className="c1"
  >
    <div
      className="c2"
      data-testid="signupform-textinput-password"
    >
      <span
        className="c3"
        data-testid="textinput-label"
      >
        Password
      </span>
      <input
        className="c4"
        data-testid="textinput-textfield"
        id="password"
        name="Password"
        onChange={[Function]}
        type="password"
      />
    </div>
  </div>
  <div
    className="c1"
  >
    <button
      className="c5"
      data-testid="signup-button-id"
      disabled={false}
      onClick={[Function]}
      type="submit"
    >
      SIGN UP
    </button>
  </div>
</div>
`;
